### ISSUE TITLE
Move `ClassStructure` to `Style`

### ISSUE THREAD

**Issue by nevans** [rubocop/rubocop#11906]
## Is your feature request related to a problem? Please describe.

With one significant exception, `--fix-layout` behaves the way we might expect a code formatter to behave: "Layout cops inspect your code for consistent use of indentation, alignment, and white space."  The entire `Layout` department will only make whitespace changes when auto-formatting, except for one cop _(or are there others?)_.  Likewise, there is only one cop in the `Layout` department that is unsafe and could potentially change the semantics of the code.

Likewise, "Style cops check for stylistic consistency of your code" is a perfect description of what the `ClassStructure` cop is doing.

Related: #11904

[unsafe]: https://github.com/rubocop/rubocop/pull/11904

## Describe the solution you'd like

Move `Layout/ClassStructure` to `Style/ClassStructure`

## Describe alternatives you've considered

If only constant literals, and `def`/`def.self` method definitions were reordered, and if they were only between "macros", then `ClassStructure` could be safe.  However it still doesn't fit in with the "purely syntax formatting" nature of the other layout cops.  Even if it were perfectly safe, it should still be moved (in my opinion).

**Comment by bbatsov** [comment]
Yeah, it's a bit tricky to classify this cop, but I prefer for it to stay as part of the Layout department, mostly because it's about the layout of a class. If I were naming it today it'd be something like `ClassLayout` or just `Class`, given it's already `Layout/name`. 

I acknowledge it's more complex than many of the cops in the department, but this doesn't make it fundamentally different. Classes should have consistent layout (structure) in the end of the day. The implementation of this cop is a different topic altogether - there's a lot of room for improvement there and I recall they were some suggestions here and there, but we never made much headway in this direction.

**Comment by nevans** [comment]
@bbatsov That makes sense, and thanks for considering my request.  And yeah, I _love_ this cop, even though I only have it enabled for a fraction of the files in my various codebases.  And I can work around my issue by locally configuring it as unsafe (#11904).  Even in that other ticket, I tried to distinguish between "unsafe because of current implementation" vs "unsafe because it simply isn't possible to be truly safe".

But perhaps I should have started with my _primary_ motivation, and that might have led to a better feature request.  What I really want is an option to only execute the ultra-safe and completely unsurprising "formatting" cops.  To me, that means only running cops where the autocorrections will not change the AST.  With this one exception, I _think_ that all of the layout cops fit this criteria, so `rubocop -x` _almost_ does what I want.

_However_, when using `rubocop -x` as a formatter there are also several `Style` cops that I wish ran as well.  E.g: double vs single quote string literals (uninterpolated), semicolon vs `then` keyword, trailing commas, and others result in the same exact AST (verified with `ruby --dump=parsetree`).  And, there are cops from other gems that I'd like to run under `-x` as well, e.g: `RSpec/EmptyLineAfterSubject`.  But I don't think those should be moved from `Style` or `RSpec` or `Rails` (etc) into `Layout`.

So, I suppose what I _really_ want is a attribute on cops that means that "autocorrections will only format the code, while leaving the AST unchanged".  And (perhaps) that attribute could default to `true` for `Layout` cops and `false` for all others.  That seemed like a bigger change to suggest than just moving this cop, but... what do you think?

Thinking through this a bit more, the popular "format on save" use-case provides good examples why its useful to have `rubocop -x` (or `rubocop --format` or whatever) as distinct from reordering code or other `rubocop -a` autocorrections, even when they are semantically identical, even when they compile into the same iseqs.  For example: even a perfectly safe reordering of methods can be incredibly confusing when you're in the middle of working on something and your code simply "disappears" (is moved to elsewhere in the file).  And converting from a ternary into if/then/else or vice versa can have the same iseq but _does_ change the grammar in a way that a formatter shouldn't.  The formatter may know that the two are semantically identical, but it doesn't know that I'm halfway through refactoring from error objects to exceptions (from `cond? ? ok : err` to `if cond? then ok else raise Err end`) or vice versa or perhaps I'm about to add an `elsif` clause and I wanted to run the tests first.  In other words, even when `rubocop -a` is _perfectly_ safe, it still isn't safe _as a formatter_.

What do you think?

**Comment by koic** [comment]
Looking at #11904, as already discussed, the `Layout` department should conform to the formatting that operates with `rubocop -x`. This means that `ClassStructure` cop, which may change the AST and makes unsafe modifications, might be better suited for the `Style` department.

This proposed department change seems to align with the unsafe autocorrect in #11904.

**Comment by bbatsov** [comment]
Yeah, I get this perspective as well. In theory the cop mostly re-arranges stuff in the class, but metaprogramming can mean that some of this reordering is unsafe.
